improve support for zones and reduce crashes

nitaybz opened the pull request:
## Status
READY

## Description
- Added support for Zone 2 and 3 in data types received by the library
- Removed the 100 limit of volume since it can get higher than that on some models
- Added STATUS commands for POWER/VOLUME/MUTE commands for Zone 2 & 3
- Added all sound modes possibilities so the app want crash on an unknown mode
- Adjust API commands and added `zone` variable to set/get specific zone settings
- power ON device when changing volume or mute state (to not return error)
- Adjusted README with the changes

nitaybz commented:
Ok so I passed all tests except the coverall again...

jupe commented on README.md:
```diff
@@ -115,15 +115,35 @@ Sending raw command:
+<Promise> onkyo.isOff(<zone>)             // resolves false if powers off
+<Promise> onkyo.pwrToggle(<zone>)         // toggle power, resolves when ready
+<Promise> onkyo.pwrOn(<zone>)             // power on, resolves when ready
+<Promise> onkyo.pwrOff(<zone>)            // power off, resolves when ready
```
this actually resolves true. Seems that it was wrong in first place so can be updated in another PR..

nitaybz replied:
fixed

jupe commented on README.md:
````diff
@@ -115,15 +115,35 @@ Sending raw command:
+<Promise> onkyo.getSoundMode(<zone>)         // resolves current sound mode
+<Promise> onkyo.setSoundMode(<mode>, <zone>) // sound mode selection, resolves when ready
 ```
+\* `<zone>` can be empty for main zone or `"zone2"`/`"zone3"`
````
propose to change something like "`<zone>` is optional and by default control main zone." . So it's more clear that this change doesn't break api usage.

nitaybz replied:
done... let me know what you think about it now

jupe commented on lib/Onkyo.js:
```diff
@@ -220,8 +238,10 @@ class Onkyo extends EventEmitter {
     const error = new OnkyoError(`Unknown data: ${encodeURIComponent(lookup)}`);
-    return Promise.reject(error);
+    this.logger.debug(error);
+    return Promise.resolve({group, data: command});
```
I'm interested to know more reasons why this resolves instead of reject ? Does this cause some problems when integrating to other projects ?

nitaybz replied:
this is an issue as you did not collect all possible data types that the receiver can send... and we don't want to crash the process because the AVR is sending extra data we didn't expect like "album artwork", "track time" and so on... and there are many that are not included in this library. most of them can be found here:
https://github.com/estbeetoo/eiscp.js/blob/master/eiscp-commands.json

jupe replied:
Originally I collect all possible data types from Onkyo manual - obviously there is more data types in other (new?) device models - which of course could be added here as well. 

> we don't want to crash the process because the AVR is sending extra data we didn't expect

There is easy way to avoid crashing, see example:
https://github.com/jupe/onkyo.js/blob/efbc0fc3a5ce39a56f885f28f22de45bc6a27e05/sample/control.js#L7

Could this approach be used instead of accepting unknown messages ?

nitaybz replied:
I feel like those errors are not really errors and should not be marked as one... this is only data type that the lib doesn't know... if you'll check the link I provided it contain much wider list and it's updated for 2013 and not necessarily for newer models... so either we include every single bit of expected result... or we allow unfamiliar data type and we just don't handle them... 
we can maybe look for a format: must be above 4 chars for example so it will include group and data

nitaybz replied:
I have to add that I've handled and added all necessary data types to control everything that the library suggest to control... so it won't cause an error to just ignore unknown data and warn about it.

jupe replied:
> I feel like those errors are not really errors and should not be marked as one

you are right, those are not error even lib can't parse them.  Maybe it should resolve something else than plain json -  e.g. instances of classes `Message` or `UnknownMessage` which holds decoded message if possible and original bytes.. I didn't like that it resolves now plain json without clear indicate that lib didn't was able to decode payload. Any thoughts ? Consumer application should know exactly if resolved data is supported and properly decoded or unknown, right ?

> we include every single bit of expected result.

I think that's what what we should do anyway.

jupe replied:
Propose to cut this part to own PR since it leads to breaking change.

nitaybz replied:
best case scenario is to extract all possible data types... do you have any idea how to quickly extract them from the link I've send and applied them to your lib in the proper format?

nitaybz replied:
it's not breaking anything if we don't handle unfamiliar data... don't you want the library to not break? at least until we manage to include all possible data types?
for me, without it, the lib is not usable as certain commands bring unexpected results... I can just add all the data types I encounter like I did before but it might still break for others and also I don't think it's best practice to have all data types in the Onkyo.js file...

nitaybz replied:
let me know how you want to proceed:
1. I'll revert this part and maybe just add the missing one for my receiver (it will probably crash on other models)
2. Maybe you can think of a way to extract all commands from the link i've sent you
3. accept the PR as it is and add all commands in the future

those are just suggestions of course and you'll decide what's best

jupe replied:
> best case scenario is to extract all possible data types

even we extract all (at the moment) possible data types it might still leads situation where new AV-amplifier introduce new protocol message that library don't understand -> library still should handle cases when unknown message is received. I think that best option is to resolve instances which is clearly understandable if message is known or unknown like I mention earlier - API could be still backward compatible. How that sounds like ?  Resolving "success" kind of JSON data can still leads unwanted behaviour in system.

> Maybe you can think of a way to extract all commands from the link i've sent you

should be rather simple to create some script that convert above json to another format, or even refactor library to support exactly same format. converting json to currently supported format is risk to break something because there is no regression tests for each payloads. Maybe worth of it anyway.

I propose that you revert this part for now because this seems to be only part from this PR that is not clear (for me) how it should be. It's more clear to create separate PR that change behaviour for this.

I'm still appreciate with your honest feedback :) This is obviously case that could be better.

jupe commented on test/onkyo.js:
```diff
@@ -140,15 +140,15 @@ describe('Onkyo', function () {
-        done();
-      });
-    });
+    // it('unrecognize', function (done) {
```
I think library should have some API for cases where unrecognised data is received - to be able to manage those properly. I'm also interested to hear reasons behind your decision :) Anyway - it's not good idea to comment out tests - instead it should be either removed totally - or refactor against new behaviour.

nitaybz replied:
I commented this so it will be your decision to refactor or remove... let me know what you think I should do... read above the reason to accept unexpected data types

nitaybz replied:
I deleted it

nitaybz commented on lib/Onkyo.js:
```diff
@@ -220,8 +238,10 @@ class Onkyo extends EventEmitter {
+    // always resolve even if data is unknown
     const error = new OnkyoError(`Unknown data: ${encodeURIComponent(lookup)}`);
-    return Promise.reject(error);
+    this.logger.debug(error);
```
changed to warn level

nitaybz commented:
So I reverted resolving with unknown data... Please accept this PR and tell me how you suggest to move forward... should we open a new PR with allowing it? do you want me to create another function that if data type is unknown it will check in the config file I sent you if it's valid?

nitaybz commented:
I've created another solution for the different data types... I've added all the commands to a file and made a funciton to check weather the current data type is available in this commands file and it will print what the data means... but it won't do anything with the data besides prevent it from returning error:
https://github.com/nitaybz/onkyo.js/tree/dataTypes

jupe commented on test/onkyo.js:
```diff
@@ -140,15 +140,6 @@ describe('Onkyo', function () {
           });
       });
     });
-    it('unrecognize', function (done) {
```
please revert this back since this functionality still exists, right ?

nitaybz replied:
Done... I reverted it

jupe reviewed (approved):
one test should be re-activated but otherwise I'm ready for merge this.

jupe commented:
> EDIT: Actually it didn't work well... My receiver is still sending data that is not even in that code base

Which (example) script you used for this ? I could try it as well. So you add all commands from [this](https://github.com/estbeetoo/eiscp.js/blob/master/eiscp-commands.json) file ? 

https://github.com/nitaybz/onkyo.js/blob/dataTypes/lib/eiscp-commands.json file format looks very good improvement for lib 👍

nitaybz commented:
> > EDIT: Actually it didn't work well... My receiver is still sending data that is not even in that code base
> 
> Which (example) script you used for this ? I could try it as well. So you add all commands from [this](https://github.com/estbeetoo/eiscp.js/blob/master/eiscp-commands.json) file ?
> 
> https://github.com/nitaybz/onkyo.js/blob/dataTypes/lib/eiscp-commands.json file format looks very good improvement for lib 👍

The file is working good and it's definitely adding more data options but unfortunately it is not covering all possibilities... still my receiver is sending unknown data...
I really believe we should resolve with a warning even if unknown data appears

nitaybz commented:
Also I've added an option for N/A state for sound mode as this is possible when the avr is off

nitaybz commented:
I improved catching errors a bit and remote control command

+ in the dataTypes branch, I updated the file to include more commands and added one manual command that I couldn't find anywhere but my AVR was sending

Now all commands are ok... but I guess it won't catch all possible new models

jupe commented on lib/Onkyo.js:
```diff
@@ -711,7 +718,6 @@ class Onkyo extends EventEmitter {
       .then(({MVL}) => MVL);
   }
   setVolume(volume, zone) {
-    invariant(_.isInteger(volume), 'volume should be an integer');
```
why to remove this check ?

nitaybz replied:
it's back... was removed by mistake when I merged

jupe commented:
one test fails:

> Onkyo api setVolume throws when invalid input - throws when invalid input

nitaybz commented:
> one test fails:
> 
> > Onkyo api setVolume throws when invalid input - throws when invalid input

fixed

nitaybz commented:
do you want me to add dataTypes branch to this pull request as well?

jupe commented:
> do you want me to add dataTypes branch to this pull request as well?

please create another PR. I would like to merge this now :)

nitaybz commented:
Please do :)
